### PR TITLE
Block Editor: Fix Multiple Trailing Inserters for Nested Inner Blocks

### PR DESCRIPTION
**Pull request by jeyip**
## Note:
This is a _potential_ fix. It seems like this issue has been discussed in the past, and there might be UX reasoning behind why we want to show duplicate trialing inserters. I think it's worthwhile to discuss, but for now, I drafted a PR to frame our conversation in something concrete. I will be posting a quick summary of alternative ways we can proceed here https://github.com/WordPress/gutenberg/issues/24360

## Description

Blocks, like the group or query loop block, render what are called [inner blocks](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/inner-blocks/index.js). These inner blocks use block list blocks to display ordered content. The trailing inserter is included as the last item in block lists.

When blocks with inner blocks are **nested** in other inner blocks, the trailing inserter is rendered multiple times. For example, if a query loop block is placed **within** a group block, both will render their own set of inner blocks, which subsequently display two trailing inserters.

One potential solution is to display a single trailing inserter at any given time. We determine which inserter to show by evaluating the currently selected block and ensuring that the trailing inserter is a sibling (that way we're certain they're children of the same inner block)

## How has this been tested?

1. Enable site editing locally
2. Navigate to the site editor
3. Click on a child of a set of inner blocks, where inner blocks are the content for group blocks, query loop blocks, etc.)
4. Verify that multiple **trailing** inserters are not displayed at any given time
5. Verify the same behavior in the post editor

### Before
Bad Trailing Inserter

### After
Trailing Inserter

## Types of changes

Potential bug fix for https://github.com/WordPress/gutenberg/issues/24360

## Checklist:
- [ ] My code is tested.
- [ ] My code follows the WordPress code style. 
- [ ] My code follows the accessibility standards. 
- [ ] My code has proper inline documentation. 
- [ ] I've included developer documentation if appropriate. 
- [x] I've updated all React Native files affected by any refactorings/renamings in this PR.

### CODE REVIEW

**Comment by noahtallen** [comment]
I like the direction this goes! It's a plus one from me, because I think showing multiple block inserters at the same level is confusing in this situation. Particularly because the two appearing close together (from the group block and from the template part block) don't _seem_ like they would insert a block into separate entities, but they actually do.

**Comment by jeyip** [comment]
### Note:

I noticed interesting behavior, which seems intended, while testing this fix. The trailing inserter is only shown if the **last** block in a set of inner blocks **is not** a paragraph. If you're not seeing the trailing inserter when you'd expect it to display, add a block that isn't a paragraph to the last position of a group block, query loop block, etc.

**Review by Addison-Stavlo — COMMENTED** [review]
This behavior seems to be working very well!  I like it and think it is an improvement.  For example if we have Group B nested in Group A, then selecting Group B will only add the appender to the end of Group B (as opposed to adding one for each nested group).  It seems like a good change because having multiple inserters show up on selection is disorienting, confusing, and unnecessary.

> I noticed interesting behavior, which seems intended, while testing this fix. The trailing inserter is only shown if the last block in a set of inner blocks is not a paragraph.

Yeah this seems to be pre-existing behavior, but I don't think it completely working as intended?  As soon as you click the appender it becomes an empty paragraph block by default and it would be odd for another appender to pop beneath that as soon as you clicked on it, so this part seems intended.  However, the way it is currently written means that appender will never appear after any paragraph block (even if it is pre-existing and full of content). 

I would think that condition that checks `if lastBlock === core/paragraph` should be expanded to also require that that last paragraph block has no actual content 🤔  (or the content starts with '/' and has no spaces so an appender isnt added when a block is being initialized via "/blockName" shortcut).  @youknowriad - do you happen to know anything about why it behaves this way?

**Comment by ZebulanStanphill** [comment]
The special-case behavior for Paragraph blocks definitely was intended at some point, but I suspect time may have changed a lot of people's opinions on it. (I recall how lots of people were on board for increasing block padding for parent blocks to make them easier to select... only for a lot of the same people to change their minds a few months later after better solutions were found and implemented. Similarly, I recall how the decision to [remove the quick-insert shortcuts](https://github.com/WordPress/gutenberg/pull/19045) was almost comically unanimous.)) Personally, I've come to think the no-appender behavior with Paragraph blocks is weird and inconsistent, since even other text-related blocks lack the behavior. I think we ought to make the behavior consistent across all blocks.

Anyway, as for this PR, I like the idea behind it. I think it's a definite improvement over `master`, keeping in the spirit of only showing controls contextually and looking like the front-end as much as possible.

I did catch one issue, though. Check out empty Column blocks:

image

**Comment by Addison-Stavlo** [comment]
> I did catch one issue, though. Check out empty Column blocks:

Ah, yes im seeing this as well.  Actually inserting the columns block is a bit uglier on my end, it ends up removing that block appender and the selected column has no height:

Screen Shot 2020-08-26 at 11 14 49 PM

Where normally we would expect:
Screen Shot 2020-08-26 at 11 15 25 PM

So this seems to be effecting more than just the sibling appender 🤔

**Comment by ZebulanStanphill** [comment]
@Addison-Stavlo This PR was never modifying the sibling inserter in the first place. It was modifying the "trailing appender" (or whatever it's called). As it turns out, the Column block placeholders are just restyled trailing appenders, hence why they're affected.

**Comment by jasmussen** [comment]
Thanks very much for tackling this head on! The issue with multiple inserters is especially frustrating in patterns where you might see a group with a background, a columns inside, then an image, and you'll end up with 3 black plus buttons. This is an important problem to solve, as in many cases it actuallly breaks the layout and definitely visually misleads.

The precise behavior is hard to get right, though. The primary challenge the trailing appender is meant to solve is the situation where a **non-text block** is the last block in a container, such as an image. While you can always select an image and press Enter when it has focus and it creates a new paragraph beneath, this in practice has not been intuitive to folks. Importantly, that issue _appears_ to be addressed in this PR, and generally I would focus the most testing on ensuring that aspect is handled well. For that case, the principle "only show one at a time" seems right.

In superficial testing, the heuristic doesn't seem perfect yet for an empty post. This seems a bit bare:

Screenshot 2020-08-27 at 08 58 29

**Comment by jeyip** [comment]
@jasmussen @ZebulanStanphill @Addison-Stavlo 

Thanks for the feedback! The columns block and empty post state should be fixed.

An important change to note is that, with the columns block, only one trailing inserter is displayed at any given time. Given the modifications we're making in this PR, this is expected, but it definitely changes the columns block experience.  I'm not sure if anyone feels strongly about UX updates.

### Before
Screen Shot 2020-08-27 at 9 43 35 PM

### After
Screen Shot 2020-08-27 at 9 43 49 PM

**Comment by Addison-Stavlo** [comment]
Nice updates, I like how this is taking shape.  However, I am noticing another side effect here.  Previously, when we had the Column itself selected we were given the appender at the bottom of that column.  Now with the column selected there doesn't seem to be an appender for it, we must select a block in that column in order to get the appender to appear:

**Before**
Screen Shot 2020-08-28 at 1 25 00 PM

**After**
Screen Shot 2020-08-28 at 1 26 00 PM
 
But still appears when the block in the column is selected:
Screen Shot 2020-08-28 at 1 27 27 PM

Regarding the 3 trailing appenders (one for each column) - This is actually behaving oddly for me on master.  I reproduced it with the 3 appender once, but am unable to reproduce it anymore.  A lot of the time I am just seeing the appender for just the one column that is selected anyways.  Even chalking this up to my own local env weirdness, and assuming that we always currently have the 3 appenders on master, I feel only showing the appender for the selected column is not a bad change.

**Comment by jeyip** [comment]
> However, I am noticing another side effect here. Previously, when we had the Column itself selected we were given the appender at the bottom of that column. Now with the column selected there doesn't seem to be an appender for it, we must select a block in that column in order to get the appender to appear

@Addison-Stavlo This is a good observation 👍. I wanted to note that this behavior _is_ expected in the sense that, in our code, we've outlined logic to _only_ display the trailing inserter for one set of inner blocks at any given time. 

To elaborate, clicking on an inner block for one column, like a Site Logo block, displays the trailing inserter for column inner blocks. Clicking on an inner block for columns, however, like one of the column blocks, should display a **different** trailing inserter for columns inner blocks (there is no trailing inserter for columns inner blocks though, which is behavior exists on master).

| Click on column inner block | Clicking on columns inner block |
| -----  | ----- |
| Screen Shot 2020-08-28 at 1 27 27 PM | Screen Shot 2020-08-28 at 1 26 00 PM |

With all of that being said, I do agree that the "Before" / previous behavior seems more ideal for user interactions. I only wanted to call this out because the change to address this would be more of an exception to our new business logic for trailing inserters, and less of a "fix" for aberrant behavior.

**Comment by noahtallen** [comment]
> Hopefully that makes sense. I'm leaning towards implementing the code to create an exception for column blocks, but let me know if you think otherwise.

Are there other blocks to which this could apply, or maybe a more general way to improve this? I guess, what exactly makes the columns block special in this situation?

**Comment by Addison-Stavlo** [comment]
> Hopefully that makes sense.

Yeah, that makes sense!  

>  I'm leaning towards implementing the code to create an exception for column blocks, but let me know if you think otherwise.

It seems to be a change to group blocks as well.  Previously with the group block selected the appender was added to the bottom inside the group block, similarly now it is not.  Im not sure what the best approach is here, but creating case by case exceptions may not be the best way to go about it. 🤔   Maybe the extra qualifier should be that if it is a block that can contain inner blocks it will still need the appender inside of it?  May need more external input here...

**Comment by jeyip** [comment]
> It seems to be a change to group blocks as well. Previously with the group block selected the appender was added to the bottom inside the group block, similarly now it is not.

Oh good catch. This is going to happen with _any_ block that implements inner blocks.

> Maybe the extra qualifier should be that if it is a block that can contain inner blocks it will still need the appender inside of it?

This would reintroduce the look of "duplicated" inserters. Imagine a nesting like group 1 > group 2 > paragraph. Group 1 and Group 2 can _both_ have inner blocks. With that logic, if group 2 is selected, an inserter would be displayed alongside the group 2 block and the paragraph block. 

> maybe a more general way to improve this?

I'm going to give the business logic more thought. My intuition says that it might not be possible with our current approach.

**Comment by jeyip** [comment]
@Addison-Stavlo I pushed changes to support the use case where we want the inserter block displayed when a parent is selected.

It reintroduces some instances where we see "duplicate" inserters, specifically when we select a block with inner blocks nested within another block. (Ex. Group 1 > Group 2 > Paragraph, where we select Group 2) I still think that overall, it's an improvement over the behavior we were struggling with before.

File: packages/block-editor/src/components/block-list-appender/index.js [hunk 1]
```diff
@@ -40,6 +41,20 @@ function BlockListAppender( {
 	} else if ( canInsertDefaultBlock ) {
 		// Render the default block appender when renderAppender has not been
 		// provided and the context supports use of the default appender.
+		const isDocumentAppender = ! rootClientId;
+		const isParentSelected = selectedBlockClientId !== rootClientId;
```
Comment (Addison-Stavlo) [inline packages/block-editor/src/components/block-list-appender/index.js:45]:
```suggestion
		const isParentSelected = selectedBlockClientId === rootClientId;
```

**Review by Addison-Stavlo — COMMENTED** [review]
>  It should address the issues you brought up with the group and column blocks.
> It reintroduces some instances where we see "duplicate" inserters

Hrm, testing seems to reintroduce the duplicates as they were without fixing the parent selection.. but I think its just a small slip in the logic (listed below).

With that change, you are right... it does fix the issue I noted with the parent selected.  It reintroduces the duplicate in some cases as you mentioned, but I agree it is still better behavior than before!

This is tricky, but it is starting to look pretty good. 😁

File: packages/block-editor/src/components/block-list-appender/index.js (hunk 1 above)
Comment (jeyip) [reply packages/block-editor/src/components/block-list-appender/index.js:45]: Ah good catch! Thanks Addie 🙌 I must have reverted it back before committing.

**Comment by jeyip** [comment]
Everything looks to be in a good place. This PR is ready for another round of CR and feedback @ZebulanStanphill @MichaelArestad @noahtallen @Addison-Stavlo 🙂

Thanks for the help everyone!

**Comment by noahtallen** [comment]
I'm not exactly sure which edge cases to test for here, so I decided to give it a spin and see how it felt. Unfortunately, I don't think this fixes the use cases I was expecting it to fix!

This is on a fresh local install. I've made no edits: just selected to the left of where the top template part is so that it would select the blocks there.

Screen Shot 2020-09-02 at 3 13 24 PM

I think the block structure in this theme is:

```
- group
  - template part
    - group (this is the one selected in the screenshot above
```

**Comment by jeyip** [comment]
> I'm not exactly sure which edge cases to test for here, so I decided to give it a spin and see how it felt. 

@noahtallen I can clarify what we can expect from this PR now:

When we select blocks with no inner blocks, there should no longer be duplicate inserters. (Ex. Selecting Paragraph from Group > Template Part > Paragraph)

Screen Shot 2020-09-02 at 5 54 58 PM

---

2. When there are sets of nested inner blocks and we select the outermost parent, there should no longer be duplicate inserters (Ex. Selecting Group from Group > Template Part > Paragraph)

Screen Shot 2020-09-02 at 5 55 25 PM

---

3. A trailing inserter for the root document will always be shown as it is today.

Screen Shot 2020-09-02 at 5 58 24 PM

**Comment by jeyip** [comment]
Seeing duplicated inserters when selecting a block, with its own inner blocks is expected. The reason is because, as previously discussed [here](https://github.com/WordPress/gutenberg/pull/24836#issuecomment-683142556), there are certain blocks (like the column and group block) where we'd like to display the trailing inserter when:

1. The block itself is selected
2. The block's children are selected

When we have a series of nested inner blocks, we run into the situation where we display 1 inserter for the selected block (because it's selected) and 1 inserter for its parent (because the parent's children are selected).

**Review by noahtallen — APPROVED** [review]
Block structure is `Group -> Template Part -> paragraph`

- ✅ Only one inserter when group is selected (for the group block)
- ❓ Double inserters when template part is selected for both group and template part.
- ✅ Only one inserter when the paragraph block is selected (for the template part block).

I think the second scenario is the exception you described? I'm not super happy with it here, but I can definitely accept that it is necessary for supporting other blocks. I do wish we could be a bit smarter about it.

That said, from my perspective looking at the template parts, this is an improvement in several cases, so I'll give it a +1! 🚀

**Comment by jeyip** [comment]
> I'm not super happy with it here, but I can definitely accept that it is necessary for supporting other blocks. I do wish we could be a bit smarter about it.

It definitely felt like we _could_ be smarter about it. I spent some time last night thinking about a few last changes we could make to improve the experience. Long story short, I add a condition that checks for nested inner blocks. It's an easy revert, but I think it noticeably improves the editing experience. Let me know what you think  @noahtallen @Addison-Stavlo

**Comment by noahtallen** [comment]
Hm. I think we might want to play it safe and not include that change for now. It is more difficult to add blocks in scenarios where you have column -> group. You can't add another block to the column unless you select the column itself in this scenario, which is very difficult if you aren't comfortable with "select parent block" or the breadcrumb.

But I do really like it for the template part use cases we are looking at :D

**Pull request merged**